### PR TITLE
feat: tile metadata table

### PR DESCRIPTION
**Pull request by blacha**
Store tile metadata inside of dynamodb so imagery records do no need to be hard coded.

### CODE REVIEW

File: packages/lambda-shared/src/aws/tile.metadata.table.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    /** Max zoom to show the layer @default 32 */
+    maxZoom: number;
+}
+
+export interface TileMetadataSetRecord extends BaseDynamoTable {
+    /** TileSet set name */
+    name: string;
+
+    projection: EPSG;
```
Comment (blacha) [inline packages/lambda-shared/src/aws/tile.metadata.table.ts]: TODO EPSG.Google

File: packages/lambda-xyz/src/__test__/xyz.test.ts [hunk 2]
```diff
@@ -42,7 +44,10 @@ o.spec('LambdaXyz', () => {
         Tilers.tile256.tile = tileMock as any;
         Tilers.compose256 = { compose: rasterMock } as any;
 
-        TiffUtil.getTiffsForQuadKey = () => [];
+        const tileSet = new TileSet('aerial', EPSG.Google, 'bucket');
```
Comment (blacha) [inline packages/lambda-xyz/src/__test__/xyz.test.ts]: TODO clear tileset map

File: packages/lambda-xyz/src/cli/imagery/regional/auckland.ts [hunk 3]
```diff
@@ -1,11 +1,10 @@
 import { EPSG } from '@basemaps/geo';
-import { MosaicCog } from '../../tiff.mosaic';
+import { MosaicCog } from '../mosaic.cog';
```
Comment (blacha) [inline packages/lambda-xyz/src/cli/imagery/regional/auckland.ts]: should we just delete all the imagery now? Its not really needed.

Other option is we could just spider it from the s3 folder.

File: packages/lambda-xyz/src/cli/imagery/tiff.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+// import { MosaicCog } from './tile.set';
```
Comment (blacha) [inline packages/lambda-xyz/src/cli/imagery/tiff.ts]: remove

File: packages/lambda-shared/src/aws/tile.metadata.table.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+            ans = ans.length == 0 ? batch : ans.concat(batch);
+
+            if (!res.LastEvaluatedKey) break;
+
+            params.ExclusiveStartKey = res.LastEvaluatedKey;
+        }
+
+        return ans;
+    }
```
Comment (jacott) [inline packages/lambda-shared/src/aws/tile.metadata.table.ts]: Do we still need this?

File: packages/lambda-xyz/src/cli/import.imagery.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * CLI to iterate over all imagery sets that have been defined and determine if all the COGS are present and optimized
+ */
+async function main(): Promise<void> {
+    const logger = LogConfig.get();
+
+    const db = new DynamoDB({ region: 'ap-southeast-2' });
+
+    const out = await db.getItem({ Key: { id: { S: 'abc1233' } }, TableName: Const.TileMetadata.TableName }).promise();
```
Comment (blacha) [inline packages/lambda-xyz/src/cli/import.imagery.ts]: remove

File: packages/lambda-xyz/src/tile.set.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+export class TileSet {
+    name: string;
+    projection: EPSG;
+    tileSet: TileMetadataSetRecord;
+    imagery: Map<string, TileMetadataImageryRecord>;
+    sources: Map<string, CogTiff>;
+    bucket: string;
+
+    static BasePath(record: TileMetadataImageryRecord, quadKey?: string): string {
+        const id = record.id.startsWith('im_') ? record.id.substr(3) : record.id;
```
Comment (blacha) [inline packages/lambda-xyz/src/tile.set.ts]: should we constify `im_` ? 

Possibly:
```
TableMetadata.prefix(Prefix, string)
TableMedadata.unprefix(Perfix, string)
```

File: packages/lambda-xyz/src/tile.set.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    get id(): string {
+        return `${this.name}_${this.projection}`;
+    }
+
+    async load(): Promise<void> {
+        this.tileSet = await Aws.tileMetadata.db.getTileSet(this.name, this.projection);
+        this.imagery = await Aws.tileMetadata.db.getAllImagery(this.tileSet);
+    }
+
+    *allImagery(): Generator<{ rule: TileMetadataImageRule; imagery: TileMetadataImageryRecord }> {
```
Comment (blacha) [inline packages/lambda-xyz/src/tile.set.ts]: Not sure if we really need this generator, maybe remove?

File: packages/lambda-xyz/src/cli/imagery/tiff.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+//     }
+
+//     // Resolution, highest resolution (lowest number) on top
+//     if (a.resolution != b.resolution) {
+//         return b.resolution - a.resolution;
+//     }
+
+//     // If everything is equal use the name to force a stable sort
+//     return a.basePath.localeCompare(b.basePath);
+// }
```
Comment (jacott) [inline packages/lambda-xyz/src/cli/imagery/tiff.ts]: Do you want to delete this file?

File: packages/lambda-shared/src/aws/tile.metadata.table.ts (hunk 5 above)
Comment (blacha) [reply packages/lambda-shared/src/aws/tile.metadata.table.ts]: Removed

File: packages/lambda-xyz/src/tile.set.ts (hunk 7 above)
Comment (blacha) [reply packages/lambda-xyz/src/tile.set.ts]: I went ahead and made this.

File: packages/lambda-xyz/src/tile.set.ts (hunk 8 above)
Comment (blacha) [reply packages/lambda-xyz/src/tile.set.ts]: Ill leave it for now.

File: packages/lambda-xyz/src/cli/imagery/regional/auckland.ts (hunk 3 above)
Comment (blacha) [reply packages/lambda-xyz/src/cli/imagery/regional/auckland.ts]: Decided on deprecating it now, after the next production release removing it.

**Pull request merged**